C/plugins/south/dht11/dht11.cpp

```diff
@@ -0,0 +1,145 @@
+ */
+#include <sstream>
+#include <iostream>
+#include <reading.h>
+#include <wiringPi.h>
+#include <dht11.h>
+#include <logger.h>
+
+#define MAX_SENSOR_READ_RETRIES	(100)
+
+using namespace std;
+
+/**
+ * Constructor for the dht11 "sensor"
+ *
+ * @param pin	The RPi pin to which DHT11 sensor's data pin is attached (e.g. pin 7 = GPIO 4, this is the default)
+ */
+DHT11::DHT11(unsigned int pin) : m_pin(pin)
+{
+	if ( wiringPiSetup() == -1 ) {
+		Logger::getLogger()->fatal("wiringPiSetup failed for dht11, can't proceed");
+		throw runtime_error("wiringPiSetup failed, can't proceed");
+	}
+	(void) takeReading(true);
+}
+
+/**
+ * Destructor for the dht11 "sensor"
+ */
+DHT11::~DHT11()
+{
+}
+
+/**
+ * Attempt to read temperature and humidity data from sensor
+ *
+ * @param sensorData   This buffer is filled with sensor data if read succeeds & data checksum matches
+ */
+bool DHT11::readSensorData(uint8_t sensorData[])
+{
+        uint8_t laststate       = HIGH;
+        uint8_t counter         = 0;
+        uint8_t j               = 0, i;
+        float   f; /* fahrenheit */
+        uint8_t dht11_dat[5] = {0, 0, 0, 0, 0};
+
+        /* pull pin down for 18 milliseconds */
+        pinMode( m_pin, OUTPUT );
+        digitalWrite( m_pin, LOW );
+        delay( 18 );
+        /* then pull it up for 40 microseconds */
+        digitalWrite( m_pin, HIGH );
+        delayMicroseconds( 40 );
+        /* prepare to read the pin */
+        pinMode( m_pin, INPUT );
+
+        /* detect change and read data */
+        for ( i = 0; i < MAX_TIMINGS; i++ )
+        {
+                counter = 0;
+                while ( digitalRead( m_pin ) == laststate )
+                {
+                        counter++;
+                        delayMicroseconds( 1 );
+                        if ( counter == 255 )
+                        {
+                                break;
+                        }
+                }
+                laststate = digitalRead( m_pin );
+
+                if ( counter == 255 )
+                        break;
+
+                /* ignore first 3 transitions */
+                if ( (i >= 4) && (i % 2 == 0) )
+                {
+                        /* shove each bit into the storage bytes */
+                        //dht11_dat[j / 8] <<= 1;
+                        uint8_t num = j>>3;
+                        dht11_dat[num] <<= 1;
+                        if ( counter > 30 ) /* was 16 */
+                                dht11_dat[num] |= 1;
+                        j++;
+                }
+        }
+
+        /*
+         * check the read 40 bits (8bit x 5 ) + verify checksum in the last byte
+         */
+        if ( (j >= 40) &&
+             (dht11_dat[4] == ( (dht11_dat[0] + dht11_dat[1] + dht11_dat[2] + dht11_dat[3]) & 0xFF) ) )
+        {
+		int k=0;
+		for (k=0; k<4; k++)
+			sensorData[k]=dht11_dat[k];
+		return true;
+        }else  {
+                //Logger::getLogger()->info( "Data not good, skip\n" );
```

Comment:
We should have some indication of this failure, if only on the first failed attempt after a good attempt or startup. Otherwise we get nothing and have no idea what is happening.

I note below this is called in a retry loop, so the log should be moved there. It certainly should not be commented out and left here.

Reply:
There are around 20-30% failures while reading the sensor and that is a common observation/behavior based on what I read in online discussion forums. So what is attempted here is that we just return the previous reading if current reading attempt fails. And in case of first reading, 100 tries are done to try to get good data (to fall back upon in future), if this fails, it would return 0.0% humidity and 0.0 degrees centigrade temperature. But putting an error log for initial reading attempt (and 100 retry failures) is a good idea, will add that.